### PR TITLE
Issue 150 store token from server in local storage 

### LINKED ISSUE #150: Store token from server in local storage (RAM or cookie idk) - Sign in and Sign up
## Basic Information
When the server responds with a JWT, store it somewhere and make sure it works with the Axios interceptor (which will append the token onto future requests
- No token refresh

https://blog.logrocket.com/using-localstorage-react-hooks/

### PR DESCRIPTION
**Pull request by ErikaKK**
## Change Summary
image

- Store token to local storage when sign in in `sign-in.tsx` page. 
- Changed fake API in poster page `index.tsx`, got task list from db API using axios interceptor in `/lib/api.ts`
- Moved and removed some poster page files bc they were not very organized before
- Add API call in poster page `new-task.tsx`, got error response
image

image
- Changed `form.tsx` and `input.tsx` bc #110 made some changes that affect the `sign-in.tsx`

### Change Form
**Fill this up (NA if not available). If a certain criteria is not met, can you please give a reason.**

- [x] The pull request title has an issue number
- [ ] The change works by "Smoke testing" or quick testing
- [ ] The change has tests
- [ ] The change has documentation

## Other Information
- Bc the design of db login table only has username field instead of email or mobile. I changed the login field label in `sign-in.tsx`

# Related issue

- Resolve #150

### CODE REVIEW

File: client/src/pages/sign-in.tsx [hunk 1]
```diff
@@ -61,7 +72,7 @@ export default function SignIn({ account }: { account: string }) {
             <SingleLineInput
               type="text"
               name="account"
-              label="Email or mobile"
+              label="Username"
```
Comment (SodaVolcano) [inline client/src/pages/sign-in.tsx]: Replace with Email or mobile, "username" is not usually associated with emails or mobile but actual user names like SodaVolcano, can be confusing

File: client/src/pages/sign-in.tsx [hunk 2]
```diff
@@ -1,42 +1,53 @@
...
+    // const isPhone = /^[0-9]{10}$/.test(formData.account);
 
-    if (!isEmail && !isPhone) {
-      setErrorMessage("Invalid email or phone number. Please try again.");
-      return;
-    }
+    // if (!isEmail && !isPhone) {
+    //   setErrorMessage("Invalid email or phone number. Please try again.");
+    //   return;
+    // }
```
Comment (SodaVolcano) [inline client/src/pages/sign-in.tsx]: Not sure if the check should be performed client-side or server-side. Maybe both? If client-side, feedback is very quick, and having server also check it is just to double make sure it's valid

File: client/src/pages/sign-in.tsx [hunk 3]
```diff
@@ -1,42 +1,53 @@
...
+    password,
   });
-}
+  const { token } = response.data;
+
+  // Store the token in localStorage
+  localStorage.setItem("token", token);
+
+  console.log("Login successful, token stored.");
+  console.log("Token", token);
```
Comment (SodaVolcano) [inline client/src/pages/sign-in.tsx]: Probably remove this if it's for debuggin purposes

File: client/src/lib/api.ts [hunk 4]
```diff
@@ -16,7 +16,7 @@ const axiosInstance = axios.create({
 axiosInstance.interceptors.request.use(
   (config: InternalAxiosRequestConfig<any>) => {
     const token = localStorage.getItem("token");
-    console.log("token:", token);
+    // console.log("token:", token);
```
Comment (SodaVolcano) [inline client/src/lib/api.ts]: remove

File: client/src/pages/poster/index.tsx [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+  const [tasks, setTasks] = useState<object[]>([]);
+  const [loading, setLoading] = useState(true);
+  const [error, setError] = useState<string | null>(null);
+
+  useEffect(() => {
+    const fetchUserData = async () => {
+      try {
+        const response = await axiosInstance.get("/app/tasks/");
+        const jsonResponse = response.data;
+        // console.log(jsonResponse);
```
Comment (SodaVolcano) [inline client/src/pages/poster/index.tsx]: remove if for debugging purposes

File: client/src/pages/poster/index.tsx [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+        const user_id = decoded.user_id;
+
+        const filteredTasks = jsonResponse.filter(
+          (item: any) => item.owner_id === user_id,
+        );
+
+        setTasks(filteredTasks);
+      } catch (error: unknown) {
+        if (error instanceof Error) {
+          console.error("Error fetching user data:", error.message);
```
Comment (SodaVolcano) [inline client/src/pages/poster/index.tsx]: remove, setError should show it on the page using `ErrorCallout`

File: client/src/pages/poster/index.tsx [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+          setError(error.message);
+        } else {
+          console.error("An unexpected error occurred:", error);
```
Comment (SodaVolcano) [inline client/src/pages/poster/index.tsx]: Remove, should be displayed by `ErrorCallout`

**Review by SodaVolcano — CHANGES_REQUESTED** [review]
Have some `console.log` that should be removed probably, but looks pretty good!

File: client/src/pages/sign-in.tsx [hunk 8]
```diff
@@ -1,42 +1,53 @@
...
-      } else {
-        resolve({ status: 404, message: "User and password not found" });
-      }
-    }, 1000);
+import { LocalBaseURL } from "@/lib/api";
+
+const LOGIN_URL = LocalBaseURL.concat("/app/login/");
+
+const handleLogin = async (username: string, password: string) => {
+  const response = await axios.post(LOGIN_URL, {
```
Comment (SeiKasahara) [inline client/src/pages/sign-in.tsx:15]: Should we use the `api` in the lib?

Comment (ErikaKK) [reply client/src/pages/sign-in.tsx:15]: I wrote this so I use it. It works well

**Comment by ErikaKK** [comment]
To satisfy current fields in db, have to use username to login. Needs to change later

File: client/src/pages/sign-in.tsx [hunk 9]
```diff
@@ -1,42 +1,50 @@
...
     const isPhone = /^[0-9]{10}$/.test(formData.account);
 
-    if (!isEmail && !isPhone) {
-      setErrorMessage("Invalid email or phone number. Please try again.");
-      return;
-    }
+    // if (!isEmail && !isPhone) {
+    //   setErrorMessage("Invalid email or phone number. Please try again.");
+    //   return;
+    // }
```
Comment (SodaVolcano) [inline client/src/pages/sign-in.tsx]: Feel free to remove validation if you think it should be done only on the server side. I guess we have to send the details to the server anyway so it doesn't matter if we also validate it there as well.

**Review by SodaVolcano — APPROVED** [review]
Looks pretty good, remove the validation if you want (can always go back to a previous commit if we need the regex pattern again)

File: client/src/pages/sign-in.tsx (hunk 9 above)
Comment (ErikaKK) [reply client/src/pages/sign-in.tsx]: Ok thanks now they only use username to login. I'll change later

Comment (ErikaKK) [reply client/src/pages/sign-in.tsx]: Now the login API uses username in the payload. I'll change later

**Comment by CodeWithTheDoctor** [comment]
I wasn't sure how to test this so can't review it.

What I tried: Opened localhost:3000, enetered my email and name, clicked "Finish", saw a confirmation on the console page.

But token doesn't appear in the localstorage?

image


Let me know if there's something else I should be looking for!

**Comment by ErikaKK** [comment]
> I wasn't sure how to test this so can't review it.
> 
> What I tried: Opened localhost:3000, enetered my email and name, clicked "Finish", saw a confirmation on the console page.
> 
> But token doesn't appear in the localstorage?
> 
> image
> 
> Let me know if there's something else I should be looking for!


Someone else is working on this signup page #134 
My team made the signin page. You can access `http://localhost:3000/signin-test` and click register to signup a user with  
```
email: "user1@example.com",
        password: "user123abc@#",
        username: "user1",
```
Then go to `http://localhost:3000/sign-in`. You have to use the username in the `Email or mobile` field to login bc that's how the api is designed. and then you can see the token in local storage
image

**Comment by ErikaKK** [comment]
@CodeWithTheDoctor

**Review by yunho7687 — COMMENTED** [review]
please fix the conflict : )

**Review by yunho7687 — APPROVED** [review]

**Pull request merged**